server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollector.java

```diff
@@ -0,0 +1,322 @@
+                    // wants to skip low-scoring hits, the aggs collector still sees all hits.
+                    // this is important also for terminate_after in case used when total hits tracking is early terminated.
+                }
+            };
+            if (topDocsLeafCollector != null) {
+                topDocsLeafCollector.setScorer(scorer);
+            }
+            if (aggsLeafCollector != null) {
+                aggsLeafCollector.setScorer(scorer);
+            }
+            this.scorer = scorer;
+        }
+
+        @Override
+        public void collect(int doc) throws IOException {
+            if (shouldCollectTopDocs(doc, scorer, postFilterBits)) {
+                if (topDocsLeafCollector != null) {
+                    try {
+                        topDocsLeafCollector.collect(doc);
+                    } catch (@SuppressWarnings("unused") CollectionTerminatedException e) {
+                        topDocsLeafCollector = null;
+                        // top docs collector does not need this segment, but the aggs collector may.
+                        if (aggsLeafCollector == null) {
+                            // TODO we keep on collecting although we have nothing to collect (there is no top docs nor aggs leaf
+                            // collector).
+                            // The reason is only to set the early terminated flag to the QueryResult. We should fix this.
+                            if (terminateAfter == 0) {
+                                throw e;
+                            }
+                        }
+                    }
+                }
+            }
+            // min_score is applied to aggs as well as top hits
+            if (minScore == null || scorer.score() >= minScore) {
+                if (aggsLeafCollector != null) {
+                    try {
+                        aggsLeafCollector.collect(doc);
+                    } catch (@SuppressWarnings("unused") CollectionTerminatedException e) {
+                        aggsLeafCollector = null;
+                        // aggs collector does not need this segment, but the top docs collector may.
+                        if (topDocsLeafCollector == null) {
+                            // TODO we keep on collecting although we have nothing to collect (there is no top docs nor aggs leaf
+                            // collector).
+                            // The reason is only to set the early terminated flag to the QueryResult. We should fix this.
+                            if (terminateAfter == 0) {
+                                throw e;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        @Override
+        public DocIdSetIterator competitiveIterator() throws IOException {
+            // we expose the competitive iterator only when one of the two sub-leaf collectors has early terminated
```

Comment:
maybe leave a TODO to consider making this a disjunction

Reply:
++